mongodb/vibe/db/mongo/connection.d

```diff
@@ -306,26 +308,50 @@ final class MongoConnection {
 	{
 		scope(failure) disconnect();
 		foreach (d; documents) if (d["_id"].isNull()) d["_id"] = Bson(BsonObjectID.generate());
-		send(OpCode.Insert, -1, cast(int)flags, collection_name, documents);
+		string collection = collection_name.canFind(".") ? collection_name.split(".")[1] : collection_name;
+		Bson operation = Bson([
+			"insert": Bson(collection),
+			"$db": Bson(m_settings.database == string.init ? "admin" : m_settings.database),
+			"documents": Bson(documents)
+		]);
+		auto id = send(OpCode.Msg, -1, cast(int)0, cast(ubyte)0, operation);
+		recvReply!Bson(id, (long cursor, ReplyFlags flags, int first, int num) {
+
+		}, (size_t idx, ref Bson doc) {});
 		if (m_settings.safe) checkForError(collection_name);
 	}
 
-	void query(T)(string collection_name, QueryFlags flags, int nskip, int nret, Bson query, Bson returnFieldSelector, scope ReplyDelegate on_msg, scope DocDelegate!T on_doc)
+	void query(T)(string collection_name, QueryFlags flags, int nskip, int nret, Bson query, Bson returnFieldSelector, scope ReplyDelegate on_msg, scope DocDelegate!T on_doc, bool isAggregate = false)
 	{
 		scope(failure) disconnect();
 		flags |= m_settings.defQueryFlags;
 		int id;
-		if (returnFieldSelector.isNull)
-			id = send(OpCode.Query, -1, cast(int)flags, collection_name, nskip, nret, query);
-		else
-			id = send(OpCode.Query, -1, cast(int)flags, collection_name, nskip, nret, query, returnFieldSelector);
-		recvReply!T(id, on_msg, on_doc);
+		Bson op = query;
+		op["$db"] = Bson(m_settings.database == string.init ? "admin" : m_settings.database);
+
+		if (!returnFieldSelector.isNull)
+			op["projection"] = returnFieldSelector;
+
+		if (flags && QueryFlags.tailableCursor)
+			op["tailable"] = Bson(true);
+		if (flags && QueryFlags.awaitData)
+			op["awaitData"] = Bson(true);
+
+		Bson operation = op;
+		id = send(OpCode.Msg, -1, cast(int)0, cast(ubyte)0, operation);
```

Comment:
There are a few more query flags that are unhandled here: `slaveOk`, `oplogReplay`, `noCursorTimeout`, `exhaust`, `partial`

Also, the `collection_name` parameter is now unused.